### ISSUE TITLE
Service graph not working after pod restart

### ISSUE THREAD

**Issue by UladzimirSemiankou** [ExpediaDotCom/haystack-service-graph#115]
After node's death in production cluster and Kafka restart Service Graph stopped responding to :8080/servicegraph requests, simply hanging indefinitely. Absolutely identical configuration on staging cluster works fine after several restarts.

Cluster config:
AWS EKS cluster
3 Kafka nodes
3 Zookeeper nodes
1 node-finder pod v1.0.18 (tried also with 1.0)
1 service graph pod v1.0.18 (tried also with 1.0)

Kafka topics:
proto-spans - consumed by node-finder
graph-nodes-18 - produced by node-finder and consumed by graph-builder
service-graph-18 - should be produced by graph-builder i guess, but is empty on both envs

Tried:
adding more RAM/CPU to service-graph
fully clearing all Kafka topics
Kafka restart with topics recreation
letting service-graph and node-finder create topics automatically
changing the number of topics' partitions
changing number of graph-node's kafka connection threads and replication factor

Strange behavior:
one time after changing service-graph's produced topic to a newly created single partition topic, it started working for a whole 5 minutes before going silent again

Kafka consumer stats:
```
TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                                                                                     HOST            CLIENT-ID
graph-nodes-18  0          86133           86133           0               graph-builder-62ccfba3-d9ec-44fd-8a4f-d16bdc824834-StreamThread-2-consumer-2b5359cc-1b3d-413c-90f2-551acd042500 /10.214.12.7    graph-builder-62ccfba3-d9ec-44fd-8a4f-d16bdc824834-StreamThread-2-consumer
```

Graph-builder logs (couldn't find a way to enable debug mode):
[graph-builder.txt](https://github.com/ExpediaDotCom/haystack-service-graph/files/3426886/graph-builder.txt)

Node finder config:
[nf.txt](https://github.com/ExpediaDotCom/haystack-service-graph/files/3427048/nf.txt)

Graph-builder config:
[gb.txt](https://github.com/ExpediaDotCom/haystack-service-graph/files/3427049/gb.txt)

At least the way to enable debug logging in graph-builder would be extremely helpful.

**Comment by mchandramouli** [comment]
@UladzimirSemiankou  were you able to repro this in haystack-docker with the proto spans produced by your application?

**Comment by UladzimirSemiankou** [comment]
We are producing quite a bit of spans, and it's an enclosed environment so I probably will not be able to replicate it locally.
Also staging environment is working fine with similar spans.
Since node-finder populates its kafka topic properly, do you think it might be a spans problem?

**Comment by mchandramouli** [comment]
Output of node-finder is a simple Json. Can you check the data in graph-nodes-18.

Objects in that should be a json representation of GraphEdge (https://github.com/ExpediaDotCom/haystack-commons/blob/master/commons/src/main/scala/com/expedia/www/haystack/commons/entities/GraphEdge.scala#L28)

Serde used: https://github.com/ExpediaDotCom/haystack-commons/blob/master/commons/src/main/scala/com/expedia/www/haystack/commons/kstreams/serde/graph/GraphEdgeValueSerde.scala#L40

Regarding debug level, you can inject an environment variable named "HAYSTACK_LOG_LEVEL" with value "DEBUG" into the container :   https://github.com/ExpediaDotCom/haystack-service-graph/blob/master/graph-builder/src/main/resources/logback.xml#L24

**Comment by SVLay** [comment]
@mchandramouli graph-node-18 has a following contents:
`{"source":{"name":"gateway","tags":{}},"destination":{"name":"products","tags":{}},"operation":"GET","sourceTimestamp":1563973184181}
{"source":{"name":"dashboard-content","tags":{}},"destination":{"name":"gateway","tags":{}},"operation":"GET","sourceTimestamp":1563973186630}
{"source":{"name":"dashboard-content","tags":{}},"destination":{"name":"gateway","tags":{}},"operation":"GET","sourceTimestamp":1563973190771}
{"source":{"name":"gateway","tags":{}},"destination":{"name":"products","tags":{}},"operation":"GET","sourceTimestamp":1563973194008}
{"source":{"name":"member-dashboard-service","tags":{}},"destination":{"name":"test-service","tags":{}},"operation":"GET","sourceTimestamp":1563973195469}
{"source":{"name":"dashboard-content","tags":{}},"destination":{"name":"gateway","tags":{}},"operation":"GET","sourceTimestamp":1563973198833}
`
That's just a part of it for the reference. For the debug mode - logs vary with another healthy environment.
Non-working environment:
`2019-07-26 15:05:46:760 haystack-service-graph-graph-builder-b4945730-bb33-48e4-9329-dd78925389e2-StreamThread-1, DEBUG, org.apache.kafka.streams.processor.internals.StreamThread,  "stream-thread [haystack-service-graph-graph-builder-b4945730-bb33-48e4-9329-dd78925389e2-StreamThread-1] Committed all active tasks [] and standby tasks [] in 0ms" `
Working environment:
`2019-07-26 15:09:32:397 haystack-service-graph-graph-builder-dbe6ebc2-658c-496b-aad2-7c5cdb57f24c-StreamThread-1, DEBUG, org.apache.kafka.streams.processor.internals.StreamThread,  "stream-thread [haystack-service-graph-graph-builder-dbe6ebc2-658c-496b-aad2-7c5cdb57f24c-StreamThread-1] Committed all active tasks [0_3, 0_7] and standby tasks [] in 102ms" `
That's also just a part it, I'll attach full logs here

[functional.txt](https://github.com/ExpediaDotCom/haystack-service-graph/files/3436624/functional.txt)
[non-functional.txt](https://github.com/ExpediaDotCom/haystack-service-graph/files/3436625/non-functional.txt)

**Comment by ashishagg** [comment]
@SVLay let me share some internals of service-graph architecture to help you debug the problem.

Lets talk about data flow first:  
a1) node-finder app reads from `proto-spans` topic. This app uses in-memory buffer for stitching spans to produce an edge (json) in different topic `graph-nodes`. 
a2) node-finder also creates a topic `haystack-node-finder-metadata` for [GlobalKTable](https://kafka.apache.org/10/javadoc/org/apache/kafka/streams/kstream/GlobalKTable.html). This is used to share metadata across different node-finder instances.
a3) You can run as many instances of it depending upon your throughput and kafka partition counts

b1) graph-builder app reads edges from `graph-nodes` topic and keeps this data in [KTable](https://kafka.apache.org/10/javadoc/org/apache/kafka/streams/kstream/KTable.html). 
b2) This table is  backed up by a changelog topic that helps us to avoid data loss during crash/redeployments. One can see a topic something as `<application.id>-service-graph-changelog`. 
b3) Run as many containers here and each stores certain portion of the graph in ktable.
b4) What is inside ktable ? It is a key-value store, where key is Edge and values are the counts for every 5min window. 

When does stitching happen?
It happens at read time. Each graph-builder also runs an http server, that fans out calls to other graph-builds and build the complete graph on the calling node. 
It is important to provide `HAYSTACK_PROP_SERVICE_HOST` if you are running multiple graph-builders - required for node discovery.

```
"containers": [
          {
            "name": "graph-builder",
            "image": "expediadotcom/haystack-service-graph-graph-builder:1.0.18",
            "env": [
             {
                "name": "HAYSTACK_PROP_SERVICE_HOST",
                "valueFrom": {
                  "fieldRef": {
                    "apiVersion": "v1",
                    "fieldPath": "status.podIP"
                  }
                }
              }
...
]
```

My guess is that one container might not be sufficient to take your load. Did you check the memory/cpu in that 5 minute interval? Will it worth a try to run more graph-builder containers?

**Comment by ashishagg** [comment]
I am also sharing config for graph-builder that we use in our env. Hope it may help. 
[graph-builder-config.txt](https://github.com/ExpediaDotCom/haystack-service-graph/files/3437038/graph-builder-config.txt)

**Comment by SVLay** [comment]
@ashishagg I've tried to scale-up Service graph, purging and cleaning source kafka topics beforehand, that didn't worked. After long debugging process we've decided to roll back to specific image - 
`expediadotcom/haystack-service-graph-graph-builder:1.0@sha256:1d6772a711bb517d500c0e42442fe2160441d996499b08c90671715ee1dc71d9`, which is now working without any problems

**Comment by ashishagg** [comment]
Interesting ! This image sha points to v1.0.17. The latest graph-builder version is 1.0.18. 
And the only change that goes in this release is upgrade the jetty version. We did it for a security vulnerability reported by github scans.

https://github.com/ExpediaDotCom/haystack-service-graph/compare/1.0.18...master

Let me verify if this is a genuine issue, and we will move to more stable and secure jetty.
Thanks for reporting this issue and sharing your fix.

**Comment by ashishagg** [comment]
I guess it should fix this
https://github.com/ExpediaDotCom/haystack-service-graph/pull/116

**Comment by SVLay** [comment]
@ashishagg I assume that 6b21fdf59196b625e4cdde9282a4ddb8577463ab Image on DockerHub is the latest master branch. including your fix?

**Comment by ashishagg** [comment]
Thanks for verifying the fix at your end. We will now release this tomorrow.

**Comment by ashishagg** [comment]
We have released 1.0.19 that contains the fix. Closing this issue.
